### PR TITLE
[FLINK-3660] Measure latency and exposes them via a metric

### PR DESCRIPTION
**Pull request by rmetzger**
This commit adds the initial runtime support for measuring latency of records going through the system.

I therefore introduced a new StreamElement, called a LatencyMarker.
Similar to Watermarks, LatencyMarkers are emitted from the sources at an configured interval. The default value for the interval is 2000 ms. The emission of markers can be disabled by setting the interval to 0. LatencyMarkers can not "overtake" regular elements. This ensures that the measured latency approximates the end-to-end latency of regular stream elements.

Regular operators (excluding those participating in iterations) forward latency markers if they are not a sink.
Operators with many outputs randomly select one to forward the maker to. This ensures that every marker exists only once in the system, and that repartition steps are not causing an explosion in the number of transferred markers.
If an operator is a sink, it will maintain the last 512 latencies from each known source instance.
The min/max/mean/p50/p95/p99 of each known source is reported using a special LatencyGauge from the sink (every operator can be a sink, if it doesn't have any outputs).

This commit does not visualize the latency in the web interface.
Also, there is currently no mechanism to ensure that the system clocks are in-sync, so the latency measurements will be inaccurate if the hardware clocks are not correct.

### CODE REVIEW

**Comment by zentol** [comment]
The metrics documentations contains a list of all measured metrics, we should add this metric there as well.

**Comment by rmetzger** [comment]
Good point, thank you! I'm not sure yet if I keep the format of the metrics like this.
I find those nested HashMaps quite clumsy. Once this PR and the metrics-in-webinterface-pr is merged, I'll add the latency to the web interface as well .. Then I can make a final decision on the format.

**Comment by tillrohrmann** [comment]
Thanks for the contribution @rmetzger. I briefly skimmed over the code and noticed that `LatencyMarker` don't participate in windows. Thus, I was asking myself what latency actually means. Could you please define properly what latency measures and how this can be a useful metric for users (especially wrt windows and window functions).

**Comment by rmetzger** [comment]
I agree that the semantics are not properly defined. I'll add a section to the documentation explaining what users can expect from the `LatencyMarks`.

**Comment by rmetzger** [comment]
I got some offline feedback from @aljoscha regarding the change to implement the methods required by `OneInputStreamOperator` and `TwoInputStreamOperator` in `AbstractStreamOperator` to avoid "polluting" all the custom operators.

I'm currently working on fixing the `EventTimeWindowCheckpointingITCase`, which seems to be failing more frequently due to my change.

**Comment by rmetzger** [comment]
@aljoscha I updated the pull request, which now builds green (just rebasing fixed the issues).
I also keep metrics now per operator, not only for the sinks.

File: docs/monitoring/metrics.md [hunk 1]
```diff
@@ -475,52 +474,76 @@ Flink exposes the following system metrics:
...
   </tbody>
 </table>
 
+
+### Latency tracking
+
+Flink allows to track the latency of records traveling through the system. To enable the latency tracking
+a `latencyTrackingInterval` (in milliseconds) has to be set to a positive value in the `ExecutionConfig`.
+
+At the `latencyTrackingInterval`, the sources will periodically emit a special record, called a `LatencyMaker`.
```
Comment (zentol) [inline docs/monitoring/metrics.md]: LatencyMaker -> LatencyMarker

File: docs/monitoring/metrics.md [hunk 2]
```diff
@@ -475,52 +474,76 @@ Flink exposes the following system metrics:
...
+The marker contains a timestamp from the time when the record has been emitted at the sources.
+Latency marker can not overtake regular user records, thus if records are queuing up in front of an operator,
```
Comment (zentol) [inline docs/monitoring/metrics.md]: marker -> markers

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 3]
```diff
@@ -149,6 +154,8 @@
 
 	private LinkedHashSet<Class<?>> registeredPojoTypes = new LinkedHashSet<>();
 
+
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java]: new lines should be removed

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 4]
```diff
@@ -94,20 +102,38 @@
...
 		this.container = containingTask;
 		this.config = config;
 		String operatorName = containingTask.getEnvironment().getTaskInfo().getTaskName().split("->")[config.getChainIndex()].trim();
 		
 		this.metrics = container.getEnvironment().getMetricGroup().addOperator(operatorName);
 		this.output = new CountingOutput(output, this.metrics.counter("numRecordsOut"));
+		this.isSink = isSink;
+		Configuration taskManagerConfig = container.getEnvironment().getTaskManagerInfo().getConfiguration();
+		int historySize = taskManagerConfig.getInteger(ConfigConstants.METRICS_LATENCY_HISTORY_SIZE, ConfigConstants.DEFAULT_METRICS_LATENCY_HISTORY_SIZE);
+		if(historySize <= 0) {
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: missing space after if

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 5]
```diff
@@ -94,20 +102,38 @@
...
+			LOG.warn("{} has been set to a value below 0: {}. Using default.", ConfigConstants.METRICS_LATENCY_HISTORY_SIZE, historySize);
+			historySize = ConfigConstants.DEFAULT_METRICS_LATENCY_HISTORY_SIZE;
+		}
+
+		latencyGauge = new LatencyGauge(this.metrics, historySize, !isSink);
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: this looks a bit odd; `latencyGauge = this.metrics.gauge(new LatencyGauge(historySize, isSink), "latency")` would be more consistent to how other metrics are registered.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 6]
```diff
@@ -347,4 +521,5 @@ public void close() {
 			output.close();
 		}
 	}
+
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: unnecessary new line

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/co/CoStreamFlatMap.java [hunk 7]
```diff
@@ -83,6 +83,7 @@ public void processWatermark2(Watermark mark) throws Exception {
 		}
 	}
 
+
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/co/CoStreamFlatMap.java]: unnecessary new line

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java [hunk 8]
```diff
@@ -273,7 +276,7 @@ public void releaseOutputs() {
 
 		// now create the operator and give it the output collector to write its output to
 		OneInputStreamOperator<IN, OUT> chainedOperator = operatorConfig.getStreamOperator(userCodeClassloader);
-		chainedOperator.setup(containingTask, operatorConfig, output);
+		chainedOperator.setup(containingTask, operatorConfig, output, streamOutputs.size() == 0);
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java]: could we not set this value in the `operatorConfig` instead?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/collector/selector/DirectedOutput.java [hunk 9]
```diff
@@ -95,11 +100,17 @@ public DirectedOutput(
 
 	@Override
 	public void emitWatermark(Watermark mark) {
-		for (Output<StreamRecord<OUT>> out : allOutputs) {
+		for (Output<StreamRecord<OUT>> out: allOutputs) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/collector/selector/DirectedOutput.java]: I think the space was good here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 10]
```diff
@@ -125,6 +126,11 @@ public void processWatermark(Watermark mark) throws Exception {
 	}
 
 	@Override
+	public void processLatencyMarker(LatencyMarker latencyMarker) throws Exception {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: Why is this necessary?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 11]
```diff
@@ -322,6 +346,151 @@ public final ChainingStrategy getChainingStrategy() {
...
+		reportOrForwardLatencyMarker(latencyMarker);
+	}
+
+	public void processLatencyMarker2(LatencyMarker latencyMarker) throws Exception {
+		reportOrForwardLatencyMarker(latencyMarker);
+	}
+
+
+	protected void reportOrForwardLatencyMarker(LatencyMarker maker) {
+		if(isSink) {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: Don't all operators now keep the latency statistics?

Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: I forgot to push my latest commit, where this is fixed ;)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java (hunk 8 above)
Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java]: We would not need to set this as a value in the `StreamConfig` class, because its basically a check if `StreamConfig.getOutEdgesInOrder().size == 0`. The problem is that `getOutEdgesInOrder()` expects the user code classloader. The classloader is usually not available anymore in the operators.

**Comment by rmetzger** [comment]
Thank you for the review @zentol and @aljoscha.
I addressed your comments and rebased the code to the current master. I'm currently testing the change again locally. Once that is completed successfully, I would like to merge the change. This pull request has been open for a long time, and I don't want to rebase it again ;)

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java (hunk 8 above)
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java]: Do we need the differentiation between sinks/non-sinks anymore now that we decided to report the latency for all operators?

Comment (rmetzger) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java]: The reason why we still differentiate between those is that we don't want to keep the latency statistics for all parallel source instances at the intermediate operators.
So in the current implementation:
- Intermediate operators maintain latency statistics for each logical source (so not much data)
- Sinks maintain latency statistics for all parallel instances of each logical source (a bit more data)
The sinks maintain the additional data to allow users debugging latency issues with individual parallel instances (for example when a machine running a source has issues).

**Comment by rmetzger** [comment]
I had an offline discussion with @aljoscha about this PR, and we decided to remove the `isSink` flag from the `AbstractStreamOperator`.
I updated the PR.

**Comment by aljoscha** [comment]
This looks good to merge now! 👍

**Comment by rmetzger** [comment]
Thanks a lot for reviewing it again!
